js/gestionPresupuesto.js

```diff
@@ -1,18 +1,126 @@
+"use strict";
+
 // TODO: Crear las funciones, objetos y variables indicadas en el enunciado
 
 // TODO: Variable global
 
+let presupuesto = 0;
+let gastos = [];
+let idGasto = 0;
+
+function actualizarPresupuesto(cantidad) {
+    if(cantidad >= 0){
+        presupuesto = cantidad;
+    }else if(cantidad < 0){
+        presupuesto = -1;
+    }
 
-function actualizarPresupuesto() {
-    // TODO
+    return presupuesto;
 }
 
 function mostrarPresupuesto() {
-    // TODO
+    let mensaje = `Tu presupuesto actual es de ${presupuesto} €`
+    return mensaje;
 }
 
-function CrearGasto() {
-    // TODO
+function CrearGasto(descripcion = "No hay descripción", valor = 0, fecha = "", ...etiquetas) {
+
+    this.descripcion = descripcion;
+
+    if(valor >= 0){
+        this.valor = valor;
+    }else{
+        this.valor = 0;
+    }
+
+    if(fecha == "" || isNaN(Date.parse(fecha))){
+        this.fecha = Date.now();
+    }else{
+        this.fecha = Date.parse(fecha);
+    }
+
+    //this.etiquetas = this.anyadirEtiquetas;
```

Comment:
Mejor esta opción

Reply:
Así se comprueban duplicados

Reply:
Buenos dias,

Si he entendido bien quedaría así:

    this.etiquetas = this.anyadirEtiquetas;
    //this.etiquetas = etiquetas;

Un saludo.

Reply:
Hola,

He ejecutado el test y me sale este error: TypeError: this.anyadirEtiquetas is not a function

Un saludo.

Reply:
Hola,

    this.anyadirEtiquetas = function(...etiquetas){
        for(let etiqueta of etiquetas){
            if(this.etiquetas.includes(etiqueta) == false){
                this.etiquetas.push(etiqueta);
            }
        }
    }

    this.etiquetas = this.anyadirEtiquetas(etiquetas);
    //this.etiquetas = etiquetas;

Da este error: TypeError: Cannot read properties of undefined (reading 'includes')

Un saludo.

Reply:
Claro. Ten en cuenta que `etiquetas` es un array (la función `CrearGasto` así lo convierte a partir de los argumentos al utilizar el operador `rest`), mientras que `anyadirEtiquetas` espera una lista de argumentos. Por tanto, tienes que llamar a `anyadirEtiquetas` con una lista de argumentos. Para ello puedes utilizar el operador [spread](https://es.javascript.info/rest-parameters-spread#spread-syntax): `this.etiquetas = this.anyadirEtiquetas(...etiquetas);`.

Es un poco lioso, pero dale una vuelta a ver si te queda claro.

Reply:
Hola, 

    this.anyadirEtiquetas = function(...etiquetas){
        for(let etiqueta of etiquetas){
            if(this.etiquetas.includes(etiqueta) == false){
                this.etiquetas.push(etiqueta);
            }
        }
    }

    this.etiquetas = this.anyadirEtiquetas(...etiquetas);

Me sigue tirando fallos:

![Screenshot_3](https://user-images.githubusercontent.com/79107261/137479110-9b098a9c-76e6-4de3-9681-2d082fa8ed02.png)

Un saludo.